File: ds_5_tree/tree_5_tree_and_linkedlist/J36-m.cpp

#include <iostream>
using namespace std ;

// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() {}

    Node(int _val) {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right) {
        val = _val;
        left = _left;
        right = _right;
    }
};

class Solution {
private:
    Node* pre;
    Node* head;
public:
    Node* treeToDoublyList(Node* root) {
        if (root == NULL) return NULL;
        InorderTree(root);
        //最终要返回的是双向循环链表，所以下面将头尾连接起来
        head->left = pre;
        pre->right = head;
        return head;
    }
    //下面操作之后，根节点不是循环双向链表的头节点。
    void InorderTree(Node* root){
        if (root == NULL)
            return;
        InorderTree(root->left);
        if(pre != NULL){
            pre->right = root;
            root->left = pre;
        }
        else
            head = root;
        //root->left = pre;
        pre = root;
        InorderTree(root->right);
    }
};

File: python-algorithm/ds_5_tree/tree_5_tree_and_linkedlist/J36-m.py
# -*- coding: utf-8 -*-
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
"""


class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        if not root: return None
        pre = None
        stk = []
        cursor = root
        while stk or cursor:
            while cursor:
                stk.append(cursor)
                cursor = cursor.left
            cursor = stk.pop()
            if not pre:
                head = cursor
            else:
                pre.right = cursor
                cursor.left = pre
            pre = cursor
            cursor = cursor.right
        #  连接首尾
        head.left = pre
        pre.right = head
        return head


